### ISSUE TITLE
Encoding column names via type-level literals in the type families-based API

### ISSUE THREAD

**Issue by 0xd34df00d** [tomjaguarpaw/haskell-opaleye#425]
I'm playing around with opaleye — a great library, only positive impressions so far, especially when the code runs correctly as soon as it typechecks, thanks for that!

Regarding the design, in particular, I personally lean to the type families-based way of defining the record types. With the current approach, though, the column names (as in the arguments to the `tableField`) are decoupled from the definition of the corresponding record type. How do you reckon adding one more type parameter to `TableField` and using type-level literals to allow something along the lines of
```haskell
data Record f = Record
  { recordId :: TableField f Int SqlInt4 NN Opt "id"
  , recordValue :: TableField f String SqlText NN Req "value"
  }
```
tying the record fields and column names?

This is especially neat when a record has a dozen of fields.

If this is a route you might consider going I'd be happy to spend some time hacking on opaleye in this direction.

**Comment by ocharles** [comment]
This is the design of `rel8`: https://rel8.readthedocs.io/en/latest/tutorial.html#defining-base-tables

**Comment by 0xd34df00d** [comment]
@ocharles that looks interesting, thanks!

Too bad I just finished writing some TH helpers to generate the required instances for the type families opaleye API :) Looks like this also isn't necessary with rel8 since it already has something Generics-based.

**Comment by tomjaguarpaw** [comment]
> I'm playing around with opaleye — a great library, only positive impressions so far, especially when the code runs correctly as soon as it typechecks, thanks for that!

Cool, that's great to hear :)

I'm glad you're having success with the type families approach.  I think it's a lot more palatable than the earlier fully-polymorphic records version.

The idea to include column names in the data type definition is nice.  I'm not sure why I didn't include that in the original design.  It completely avoids the need for adaptors or explicit product-profunctor operations when defining the table.  Could you show me your design?

**Comment by 0xd34df00d** [comment]
Yeah, I also find the TF approach more strict and expressive (especially given that it's way more straightforward to add additional semantics to fields via extra type parameters to `TableField`).

Re the design, I'm not sure yet. I already have a small TH function `makeTable` that takes a couple of parameters (including the list of column names) and produces a function with the `table` invocation. My thoughts were to instead take the field names from the record field types and avoid passing them to `makeTable` explicitly.

I only have a passing familiarity with product profunctors, so I'm not sure if there's a better way. Maybe you have something better in mind?

**Comment by tomjaguarpaw** [comment]
Maybe show me your `makeTable` function and we'll go from there.
